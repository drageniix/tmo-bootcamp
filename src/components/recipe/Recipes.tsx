import React, { useContext } from "react";
import { AppContext } from "../../context/AppContext";

export const RecipeList = (): React.ReactElement => {
  const { recipes } = useContext(AppContext);

  return (
    <ul className="recipeList" data-testid="recipes">
      {recipes.map((recipe) => (
        <li key={recipe.id} className="recipeList__item">
          <img
            className="recipeList__item--img"
            src={recipe.image}
            alt="recipe preview"
          />
          <h1 className="recipeList__item--title">{recipe.title}</h1>
        </li>
      ))}
    </ul>
  );
};

export default RecipeList;
