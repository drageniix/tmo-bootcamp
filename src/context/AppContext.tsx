import React, { createContext } from "react";
import { BasicIngredient, Recipe } from "../api/recipe";
import { useRecipeSearch } from "../hooks/context";

export type AppContextType = {
  onSearchIngredients?: (input: string) => Promise<void>;
  onAddIngredient?: (chosenIngredient: BasicIngredient) => Promise<void>;
  onRemoveIngredient?: (
    id: number
  ) => (e: React.MouseEvent<HTMLButtonElement, MouseEvent>) => Promise<void>;
  ingredients: BasicIngredient[];
  searchedIngredients: BasicIngredient[];
  recipes: Recipe[];
};

export const defaultContext: AppContextType = {
  ingredients: [],
  recipes: [],
  searchedIngredients: [],
};

export const AppContext = createContext(defaultContext);

export const AppProvider = ({
  children,
}: {
  children: JSX.Element;
}): JSX.Element => {
  const appProps = useRecipeSearch();

  return <AppContext.Provider value={appProps}>{children}</AppContext.Provider>;
};
