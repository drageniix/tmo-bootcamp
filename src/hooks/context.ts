import { useState } from "react";
import {
  BasicIngredient,
  searchIngredients,
  searchRecipes,
} from "../api/recipe";
import { AppContextType, defaultContext } from "../context/AppContext";

export const useRecipeSearch = (): AppContextType => {
  const [recipes, setRecipes] = useState(defaultContext.recipes);
  const [ingredients, setIngredients] = useState(defaultContext.ingredients);
  const [searchedIngredients, setSearchedIngredients] = useState(
    defaultContext.searchedIngredients
  );

  const onSearchIngredients = async (input: string) => {
    const ingredients = await searchIngredients(input);
    setSearchedIngredients(ingredients);
  };

  const onAddIngredient = async (chosenIngredient: BasicIngredient) => {
    if (!ingredients.includes(chosenIngredient)) {
      const updatedIngredients = [...ingredients, chosenIngredient];
      const recipes = await searchRecipes(
        updatedIngredients.map((ingredient) => ingredient.name)
      );

      setIngredients(updatedIngredients);
      setRecipes(recipes);
    }
  };

  const onRemoveIngredient = (id: number) => async (
    e: React.MouseEvent<HTMLButtonElement, MouseEvent>
  ) => {
    e.preventDefault();
    const updatedIngredients = ingredients.filter(
      (ingredient) => ingredient.id !== id
    );
    const recipes = await searchRecipes(
      updatedIngredients.map((ingredient) => ingredient.name)
    );

    setIngredients(updatedIngredients);
    setRecipes(recipes);
  };

  return {
    onSearchIngredients,
    onAddIngredient,
    onRemoveIngredient,
    searchedIngredients,
    ingredients,
    recipes,
  };
};
