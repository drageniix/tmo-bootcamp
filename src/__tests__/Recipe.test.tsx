import React from "react";
import { cleanup, render } from "@testing-library/react";
import { AppContext, AppContextType } from "../context/AppContext";
import Recipes from "../components/recipe/Recipes";
import mockRecipes from "./mocks/recipes.json";

const renderForm = (context?: AppContextType) =>
  render(
    <AppContext.Provider value={context}>
      <Recipes />
    </AppContext.Provider>
  );

describe("should test ingredient list", () => {
  afterEach(cleanup);

  it("should match list snapshot", () => {
    expect.assertions(3);

    const { getByTestId } = renderForm({
      ingredients: [],
      recipes: [mockRecipes[0], mockRecipes[1]],
      searchedIngredients: [],
    });

    const listComponent = getByTestId("recipes") as HTMLInputElement;
    expect(listComponent).toMatchSnapshot();
    expect(listComponent.textContent).toContain(mockRecipes[0].title);
    expect(listComponent.textContent).toContain(mockRecipes[1].title);
  });
});
