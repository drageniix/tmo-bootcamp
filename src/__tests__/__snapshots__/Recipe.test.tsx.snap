// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should test ingredient list should match list snapshot 1`] = `
<ul
  class="recipeList"
  data-testid="recipes"
>
  <li
    class="recipeList__item"
  >
    <img
      alt="recipe preview"
      class="recipeList__item--img"
      src="https://spoonacular.com/recipeImages/509531-312x231.jpg"
    />
    <h1
      class="recipeList__item--title"
    >
      Enchilada Chicken Mango Salad
    </h1>
  </li>
  <li
    class="recipeList__item"
  >
    <img
      alt="recipe preview"
      class="recipeList__item--img"
      src="https://spoonacular.com/recipeImages/16513-312x231.jpg"
    />
    <h1
      class="recipeList__item--title"
    >
      Chicken, Sundried Tomatoes And Lettuce
    </h1>
  </li>
</ul>
`;
